libs/types-common/src/Data/Handle.hs

```diff
@@ -23,37 +31,43 @@ newtype Handle
   deriving newtype (ToJSON, ToByteString, Hashable)
 
 instance FromByteString Handle where
-  parser = parser >>= maybe (fail "Invalid handle") return . parseHandle
+  parser = handleParser
 
 instance FromJSON Handle where
   parseJSON =
     withText "Handle" $
-      maybe (fail "Invalid handle") pure . parseHandle
+      either (fail . ("Invalid handle: " <>)) pure . parseHandleEither
 
 parseHandle :: Text -> Maybe Handle
-parseHandle t
-  | isValidHandle t = Just (Handle t)
-  | otherwise = Nothing
+parseHandle = either (const Nothing) Just . parseHandleEither
 
 isValidHandle :: Text -> Bool
-isValidHandle t =
-  either (const False) (const True) $
-    parseOnly handle t
+isValidHandle = isRight . parseHandleEither
+
+parseHandleEither :: Text -> Either String Handle
+parseHandleEither = Atto.parseOnly (handleParser <* Atto.endOfInput) . Text.E.encodeUtf8
+
+handleParser :: Atto.Parser Handle
+handleParser = do
+  bs <- matching (Atto.count 2 handleChar *> Atto.skipMany handleChar)
+  when (BS.length bs > 256) $
```

Comment:
Why not use count here like before?

Reply:
No fundamental reason. `count n (optional _)` seems like it would always spin `n` times, even if there is no valid handleChar anymore. In addition, it will allocate a bunch of unnecessary `Maybe`s. Maybe the error message is also clearer? On the other hand, `skipMany` will not stop after 256 chars, so not sure which one is better. :man_shrugging: